File: src/main/java/com/freya02/botcommands/api/localization/DefaultLocalizationTemplate.java
package com.freya02.botcommands.api.localization;

import com.freya02.botcommands.internal.localization.*;
import org.jetbrains.annotations.NotNull;

import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Default implementation for {@link LocalizationTemplate}
 * <p><b>Specification:</b>
 * <br>In a nutshell, this is a copy of the {@link MessageFormat} specification, but with named parameters
 * <br>To declare a variable inside your localization template, you may use "<code>{variable_name}</code>"
 * <br>As this supports {@link MessageFormat}, you can also specify format types, such as: "<code>{variable_name, number}</code>",
 * and format styles, such as: "<code>{@literal {user_amount, choice, 0#users|1#user|1<users}}</code>"
 *
 * <p>Full example: <code>{@literal "There are {user_amount} {user_amount, choice, 0#users|1#user|1<users} and my up-time is {uptime, number} seconds"}</code>
 */
public class DefaultLocalizationTemplate implements LocalizationTemplate {
	private static final Pattern BRACKETS_PATTERN = Pattern.compile("\\{.*?}");
	private static final Pattern TEMPLATE_PATTERN = Pattern.compile("\\{(\\w+?)(?::(%.+))?}");
	private static final Pattern MESSAGE_FORMAT_PATTERN = Pattern.compile("\\{(\\w+)(,?.*?)}");

	private final List<LocalizableString> localizableStrings = new ArrayList<>();
	private final String template;

	public DefaultLocalizationTemplate(@NotNull String template, @NotNull Locale locale) {
		this.template = template;

		final Matcher bracketMatcher = BRACKETS_PATTERN.matcher(template);

		int start = 0;
		while (bracketMatcher.find()) {
			addRawString(template.substring(start, bracketMatcher.start()));

			final Matcher templateMatcher = TEMPLATE_PATTERN.matcher(bracketMatcher.group());
			if (templateMatcher.matches()) {
				final String formatterName = templateMatcher.group(1);
				final String formatter = templateMatcher.group(2);

				localizableStrings.add(new JavaFormattableString(formatterName, formatter));
			} else {
				final Matcher messageFormatMatcher = MESSAGE_FORMAT_PATTERN.matcher(bracketMatcher.group());
				if (!messageFormatMatcher.matches()) {
					throw new IllegalArgumentException("Invalid MessageFormat format '%s' in template '%s'".formatted(bracketMatcher.group(), template));
				}

				final String formatterName = messageFormatMatcher.group(1);
				final String messageFormatter = messageFormatMatcher.replaceFirst("{0$2}"); //Replace named index by integer index

				localizableStrings.add(new MessageFormatString(formatterName, messageFormatter, locale));
			}

			start = bracketMatcher.end();
		}

		addRawString(template.substring(start));
	}

	private void addRawString(String substring) {
		if (substring.isEmpty()) return;

		localizableStrings.add(new RawString(substring));
	}

	@Override
	@NotNull
	public String localize(Localization.Entry... args) {
		final StringBuilder sb = new StringBuilder();

		for (LocalizableString localizableString : localizableStrings) {
			if (localizableString instanceof RawString rawString) {
				sb.append(rawString.get());
			} else if (localizableString instanceof FormattableString formattableString) {
				final Object value = getValueByFormatterName(args, formattableString.getFormatterName());
				try {
					sb.append(formattableString.format(value));
				} catch (Exception e) { //For example, if the user provided a string to a number format
					throw new RuntimeException("Could not get localized string from FormattableString '%s' with value '%s'".formatted(formattableString.getFormatterName(), value), e);
				}
			}
		}

		return sb.toString();
	}

	private Object getValueByFormatterName(Localization.Entry[] args, String formatterName) {
		for (Localization.Entry entry : args) {
			if (entry.key().equals(formatterName)) {
				return entry.value();
			}
		}

		throw new IllegalArgumentException("Could not find format '%s' in template: '%s'".formatted(formatterName, template));
	}

	@Override
	public String toString() {
		return template;
	}
}

File: src/main/kotlin/com/freya02/botcommands/api/localization/DefaultLocalizationTemplate.kt
package com.freya02.botcommands.api.localization

import com.freya02.botcommands.api.BContext
import com.freya02.botcommands.api.core.service.getInterfacedServices
import com.freya02.botcommands.api.core.utils.simpleNestedName
import com.freya02.botcommands.api.localization.arguments.FormattableArgument
import com.freya02.botcommands.api.localization.arguments.factories.FormattableArgumentFactory
import com.freya02.botcommands.internal.localization.LocalizableArgument
import com.freya02.botcommands.internal.localization.RawArgument
import com.freya02.botcommands.internal.localization.SimpleArgument
import com.freya02.botcommands.internal.utils.throwUser
import java.text.MessageFormat
import java.util.*

private val argumentRegex = Regex("""\{(.*?)}""")
private val alphanumericRegex = Regex("""\w+""")

/**
 * Default implementation for [LocalizationTemplate].
 *
 * This is effectively [MessageFormat], but with named parameters.
 *
 * To declare a variable inside your localization template, you may use `{variable_name}`.
 *
 * As this supports [MessageFormat], you can also specify format types, such as: `{variable_name, number}`,
 * and format styles, such as: `{user_amount, choice, 0#users|1#user|1<users}`.
 *
 * Full example: `"There are {user_amount} {user_amount, choice, 0#users|1#user|1<users} and my up-time is {uptime, number} seconds"`
 */
class DefaultLocalizationTemplate(context: BContext, private val template: String, locale: Locale) : LocalizationTemplate {
    private val localizableArguments: MutableList<LocalizableArgument> = ArrayList()

    init {
        val formattableArgumentFactories = context.getInterfacedServices<FormattableArgumentFactory>()

        var start = 0
        argumentRegex.findAll(template).forEach argumentsLoop@{ argumentMatch ->
            val matchStart = argumentMatch.range.first
            addRawArgument(template.substring(start, matchStart))

            val formattableArgument = argumentMatch.groups[1]?.value!!
            // Try to match against each factory
            formattableArgumentFactories.forEach { factory ->
                factory.regex.matchEntire(formattableArgument)?.let {
                    localizableArguments += factory.get(it, locale)
                    start = argumentMatch.range.last
                    return@argumentsLoop
                }
            }

            // If the entire thing looks like a simple argument name
            if (formattableArgument.matches(alphanumericRegex)) {
                localizableArguments += SimpleArgument(formattableArgument)
                start = argumentMatch.range.last
                return@argumentsLoop
            }

            throwUser("Could not match formattable argument '$formattableArgument' against ${formattableArgumentFactories.map { it.javaClass.simpleNestedName }}")
        }
        addRawArgument(template.substring(start))
    }

    private fun addRawArgument(substring: String) {
        if (substring.isEmpty()) return
        localizableArguments += RawArgument(substring)
    }

    override fun localize(vararg args: Localization.Entry): String {
        return localizableArguments.joinToString("") { localizableArgument ->
            when (localizableArgument) {
                is RawArgument -> localizableArgument.get()
                is FormattableArgument -> formatFormattableString(args, localizableArgument)
                else -> throwUser("Unknown localizable argument type: ${localizableArgument::class.simpleNestedName}")
            }
        }
    }

    private fun formatFormattableString(args: Array<out Localization.Entry>, formattableArgument: FormattableArgument): String {
        val value = getValueByFormatterName(args, formattableArgument.formatterName)
        return try {
            formattableArgument.format(value)
        } catch (e: Exception) { //For example, if the user provided a string to a number format
            throw RuntimeException("Could not get localized string from ${formattableArgument::class.simpleNestedName} '${formattableArgument.formatterName}' with value '$value'", e)
        }
    }

    private fun getValueByFormatterName(args: Array<out Localization.Entry>, formatterName: String): Any {
        return args.find { it.key == formatterName }
            ?: throw IllegalArgumentException("Could not find format '$formatterName' in template: '$template'")
    }

    override fun toString(): String = template
}
